blockchain/blockindex.go

```diff
@@ -598,12 +613,59 @@ func (bi *blockIndex) HaveBlock(hash *chainhash.Hash) bool {
 	return hasBlock
 }
 
+// shortBlockKey generates a short identifier from a standard block hash for use
+// as a key in the block index.
+func shortBlockKey(hash *chainhash.Hash) uint32 {
+	// Use the first bytes of the hash directly since it is the result of a hash
+	// function that produces a uniformly-random distribution.  It is also worth
+	// noting that the mining process reduces entropy by zeroing the bits at the
+	// other end of the array, but there would need to be effectively impossible
+	// to achieve hash rates exceeding ~2^215.77 hashes/sec (aka ~89.8 peta
+	// yotta yotta hashes/sec) in order to start zeroing out the bits used here.
+	// Not only is that impossible for all intents and purposes, the only effect
+	// would be reducing overall memory savings due to increased collisions
+	// among the shortened keys.
+	return binary.BigEndian.Uint32(hash[0:4])
+}
+
 // addNode adds the provided node to the block index.  Duplicate entries are not
 // checked so it is up to caller to avoid adding them.
 //
 // This function MUST be called with the block index lock held (for writes).
 func (bi *blockIndex) addNode(node *blockNode) {
-	bi.index[node.hash] = node
+	// Add the node to the block index based on a shortened key with proper
+	// collision handling.  Two maps are used to that end:
+	// 1) An index map keyed by the shortened key
+	// 2) A collisions map keyed by the full hash
+	//
+	// The entry in the index map is either a block node directly when there has
+	// not been any collisions for the shortened key, or nil when there has.
+	// The nil entry signals the block node is stored in the collisions map
+	// keyed by its full hash instead.
+	blockKey := shortBlockKey(&node.hash)
+	existingNode, ok := bi.index[blockKey]
+	switch {
+	// Add the new node directly to the index when there is no collision with
+	// the short key (aka there is no existing entry for it).
+	case !ok:
+		bi.index[blockKey] = node
+
+	// Add the new node to the collision map using the full hash as the key when
+	// there is a collision with the short key and the existing node has already
+	// been moved to the collisions map implying there was a previous collision.
+	case existingNode == nil:
+		bi.collisions[node.hash] = node
+
+	// At this point, there is an existing node for the shortened key indicating
+	// there were no previous collisions for it, so move that existing node to
+	// the collision map and nil out the entry in the index map to indicate the
+	// collision.  Finally, add the new node to the collision map using its full
+	// hash as the key as well.
+	case existingNode.hash != node.hash:
+		bi.index[blockKey] = nil
+		bi.collisions[existingNode.hash] = existingNode
+		bi.collisions[node.hash] = node
+	}
```

Comment:
If hash is the same, this will keep the old node as apposed to replacing with the new node as was the previous behavior. I don't know if that matters.

Reply:
Correct.  I wasn't planning on retaining that behavior because there are many (extremely reasonable) implicit assumptions in several places that collisions are impossible in the hash space of 2^256 (this is true for all blockchains, not just DCR, and blockchain security as a whole even relies on it), so it's realistically a case that is, for all intents and purposes, impossible to hit.  

However, I suppose the existing behavior probably should be retained for testing purposes as we might want to intentionally replace a node and it would likely be surprising that it would behave differently than standard maps.

I went ahead and updated it accordingly and also added a full suite of tests which thoroughly exercise the various semantics, including replacements.